### ISSUE TITLE
clang-format: Inconsistent single-line lambda formatting

### ISSUE THREAD

**Issue by 151eb9fe-500c-4e05-b81b-4623c4c4ca20** [llvm/llvm-project#50274]
|  |  |
| --- | --- |
| Bugzilla Link | [50930](https://llvm.org/bz50930) |
| Version | 12.0 |
| OS | Linux |

## Extended Description 
Adding an empty line inside a lambda body changes the output of clang-format.

Feeding the output of clang-format into clang-format again produces a different result.

---
.clang-format:
```
BreakBeforeBraces: Custom
BraceWrapping:
  BeforeLambdaBody: true
```

---
test.cpp:
```
int main() {
  std::vector<int> v;
  std::find_if(v.begin(), v.end(),
               []
               {
                 return 1;

               });
}
```

---
clang-format test.cpp:
```
int main() {
  std::vector<int> v;
  std::find_if(v.begin(), v.end(),
               []
               {
                 return 1;
               });
}
```

---
clang-format test.cpp | clang-format:
```
int main() {
  std::vector<int> v;
  std::find_if(v.begin(), v.end(), [] { return 1; });
}
```

**Comment by mkurdej** [comment]
I don't think, this is a bug. It's just that clang-format needs two runs to get to this result.

With input:
```
auto l = []
{
  return;

};
```

In the first run, clang-format removes the empty line producing:
```
auto l = []
{
  return;
};
```

Then, on the second run, as `AllowShortLambdasOnASingleLine` is equal to `All` in default LLVM style (not overridden by your .clang-format config), clang-format now considers it as a short lambda and puts on the same line:
```
auto l = [] { return; };
```

@mydeveloperday, WDYT?

**Comment by mydeveloperday** [comment]
I agree this isn't a bug, this comes up from time to time that people think clang-format should do it all in one go, but I never really saw that as a requirement especially if you are using clang-format with the "Format on Save" option, most people are saving all the time, the blank line would be removed almost immediately after it was added for most users in my view.

I also think the fact that some people are waiting until the CI system says there is a formatting bug, also shows the wrong usage of clang-format. If people are so averse to running it integrated as part of their IDE/workflow then I don't really feel they are using it quite correctly. (my 2c worth)

Lets reject this issue as functioning as designed.

**Comment by mkurdej** [comment]
Closing.

**Comment by kalinnikovv** [comment]
Is this documented somewhere? I believe if it is working as intented then it should be documented.

**Comment by mydeveloperday** [comment]
Feel free to submit a patch with a documentation note to that effect, but unless we want clang-format to be iterative and keep reprocessing until it sees no other changes we cannot guarantee that a later pass might not impact how a previous pass might have behaved. I guess my previous point is that "continuous application" for clang-format during the editing process, this becomes a non issues. 

Only those that see reformatting as a one time pass would be impacted by such a change, and so I would question if that is the correct usage. (its a usage, but is it the best usage)

Bottom line I don't think we'd be addressing this issue as and sort of priority, given our 500+ bug backlog and the very few developers working on clang-format this just isn't any type of priority.  I feel its better we are "honest" about our intentions up front if we feel we'll never get to something like this.
